proposals/0000-split-wunused-imports.md

```diff
@@ -0,0 +1,230 @@
+1. It is on by default.
+2. It **is not made into an error by `-Werror`**. Optionally there could be `-Wmeta-error` or something along those lines for users who want to get errors for potential future meta warnings.
+
+This `-Wmeta-unused-improts` warning is only present for a few release cycles, and is removed afterwards, or alternatively, turned off by default.
+
+## Examples
+
+###
+
+```haskell
+import Foo
+import Foo (x)
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Foo` import is duplicate
+
+###
+
+```haskell
+import Foo (x)
+import Bar (x)
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Bar` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Bar` import is duplicate
+
+###
+
+```haskell
+import Foo
+import Bar
+
+bla = x
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Bar` import is unused
+
+Proposed:
+* with `-Wunused-imports` - nothing
+* with `-Wduplicate-imports` - warn that the `Bar` import is duplicate
+
+###
+
+```haskell
+import Foo
+
+bla = ()
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+* with `-Wduplicate-imports` - nothing
+
+###
+
+```haskell
+import Foo (x)
+
+bla = ()
+```
+
+Current:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+
+Proposed:
+* with `-Wunused-imports` - warn that the `Foo` import is unused
+* with `-Wduplicate-imports` - nothing
+
+## Effect and Interactions
+
+Unsure what to fill in here, it seems that the [Proposed Change Specification](#proposed-change-specification) covers the effects of this change.
+
+## Costs and Drawbacks
+
+The main cost is changing the behaviour of a warning without notice, even if we explicitly warn users that it has changed.
+
+Is this acceptable?
+
+We're in the camp of "yes, it is", **if** we include the meta warning, since anyone who is maintaining code and cares about keeping things warning free should be able to
+easily notice the meta warning and turn on `-Wduplicate-imports` (or turn off the meta warning).
+
+There's also the cost of introducing the new concept of a "meta warning". These could arguably warrant their own standalone proposal, where they are more precisely specified.
+
+## Alternatives
+
+### Don't include the meta warning and let users deal with the silently changed behaviour.
+This seems potentially frustrating. Who knows if someone has some niche use case that they really like for the "duplicate imports" part of `-Wunused-imports`?
+
+### Relaxed redundant imports
+
+We could also instead implement the spec that's suggested in the ["relaxed redundant imports" proposal](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/relaxed-unused-imports), however
```

Comment:
The wiki page may look scary, with GHC code fragments etc, but here is the gist of @bgamar's idea in one sentence:

> Then, for each use of an imported name, we will attribute
that use to all import-items which implicitly provide it, AND one import-item
which explicitly provides it. Then, any import items with no
uses attributed to them are unused, and are warned about.

I find this matches my gut intuition for how things should work quite well: if I import something explicit, then there should be only one mention of it (but additional implicit imports are fine). If there are implicit imports it shouldn't matter if there is more than one.

So I'd keep it on the table for now.

Reply:
I read through that section and mostly understood what it was trying to accomplish, but regardless, in my eyes, we should strive to keep this as straightforward as possible - the language itself is already brimming with complexity required just to get started, and I would much prefer (as a user) to not have to spend extra time to understand the guard rails of the guard rails :sweat_smile: 

I would like to reiterate:
> They also seem to be mostly orthogonal to me - if someone wants to have duplicate import warnings as per the "relaxed redundant imports" spec, then we could have another proposal after this one, potentially amending the new `-Wduplicate-imports` warning instead.

Even if the relaxed version is a straight up upgrade, I don't see why these "two" functionalities are under the same flag in the first place.

(As an anecdote, I was explaining what `-Wredundant-imports` does to a non-Haskell programmer, and even in the middle of the explanation, when I got to the "duplicate" part which this proposal is about splitting off, they interrupted to ask "but that's not a redundant import, is it?")

Reply:
That's a fair point, one can indeed argue that all imports (not just the first one) are “used” in some way.